### DISCUSSION TITLE
Tests are running, but cargo-mutants detects 0 mutants to test

### DISCUSSION THREAD

**Discussion by kevin-valerio** [sourcefrog/cargo-mutants#385] [Q&A] [answered]
Hi! 

I'm struggling a bit to make cargo-mutants working. Somehow I cannot make it working in the same way as I run my tests. 
Am I missing something ? I tried various possibilities, especially [passing options to Cargo](https://mutants.rs/cargo-args.html#passing-options-to-cargo)  or [playing with workspaces](https://mutants.rs/workspaces.html), but nothing worked.

Output after running `cargo +stable mutants`:
```bash
~/a/b/c/d/myproject > cargo +stable mutants 

Found 0 mutants to test
WARN No mutants found under the active filters
```

Output after running `cargo +stable test`
```bash
~/a/b/c/d/myproject > cargo +stable test 

running 19 tests
test tests::xx::abc ... ok
test tests::xx::abcd ... ok
...
```

Thank you very much, for your help, and for the awesome tooling!  :)

**Comment by sourcefrog** [comment]
The discovered mutants depend on the source of your tree, so without knowing anything about that it's hard to say. Is there anything unusual about it? Can you reproduce this in a smaller tree that you can share?

Is there any config in `.cargo/mutants.toml`.

What does `cargo mutants --list-files` show?

Can you share the debug output from `cargo mutants --list -Ltrace`?

**Reply by kevin-valerio** [reply]
Hey Martin! 

Thanks for your support. I don't have any `.cargo/mutants.toml`

2. Interestingly, it doesn't show the files under `tests/`
```bash
kevin@big-fuzz ~/polkadot-sdk/cumulus/parachains/integration-tests/emulated/tests/bridges/bridge-hub-rococo (master)> cargo mutants --list-files
cumulus/parachains/integration-tests/emulated/tests/bridges/bridge-hub-rococo/src/lib.rs
kevin~/polkadot-sdk/cumulus/parachains/integration-tests/emulated/tests/bridges/bridge-hub-rococo (master)> ll src/
total 8.0K
-rw-rw-r-- 1 kevin 2.9K Jul 16 11:38 lib.rs
drwxrwxr-x 2 kevin 4.0K Jul 16 11:38 tests/
kevin@big-fuzz ~/polkadot-sdk/cumulus/parachains/integration-tests/emulated/tests/bridges/bridge-hub-rococo (master)> ll src/tests/
total 68K
-rw-rw-r-- 1 kevin fuzz  21K Jul 16 11:38 asset_transfers.rs
-rw-rw-r-- 1 kevin fuzz 6.1K Jul 16 11:38 mod.rs
-rw-rw-r-- 1 kevin fuzz 4.7K Jul 16 11:38 send_xcm.rs
-rw-rw-r-- 1 kevin fuzz  22K Jul 16 11:38 snowbridge.rs
-rw-rw-r-- 1 kevin fuzz 1019 Jul 16 11:38 teleport.rs
```

3. For `cargo mutants --list -Ltrace` :
```bash
kevin@big-fuzz ~/polkadot-sdk/cumulus/parachains/integration-tests/emulated/tests/bridges/bridge-hub-rococo (master)> cargo mutants --list -Ltrace
DEBUG get_command_output{argv=["/home/kevin/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/cargo", "locate-project", "--workspace"]}: output: {"root":"/home/kevin/polkadot-sdk/Cargo.toml"}
DEBUG Found workspace root manifest cargo_toml_path="/home/kevin/polkadot-sdk/Cargo.toml"
DEBUG Find root files manifest_path="/home/kevin/polkadot-sdk/Cargo.toml" dir="/home/kevin/polkadot-sdk"
DEBUG Found workspace root workspace_root="/home/kevin/polkadot-sdk"
DEBUG config=Config { error_values: [], examine_globs: [], exclude_globs: [], exclude_re: [], examine_re: [], additional_cargo_args: [], additional_cargo_test_args: [], minimum_test_timeout: None, test_tool: None, timeout_multiplier: None, build_timeout_multiplier: None }
DEBUG args.features=Features { features: [], no_default_features: false, all_features: false }
DEBUG options=Options { baseline: Run, check_only: false, gitignore: true, in_place: false, leak_dirs: false, test_timeout: None, test_timeout_multiplier: None, build_timeout: None, build_timeout_multiplier: None, minimum_test_timeout: 20s, print_caught: false, print_unviable: false, show_times: true, show_all_logs: false, show_line_col: true, shuffle: true, additional_cargo_args: [], additional_cargo_test_args: [], features: Features { features: [], no_default_features: false, all_features: false }, examine_globset: None, exclude_globset: None, examine_names: RegexSet([]), exclude_names: RegexSet([]), output_in_dir: None, jobs: None, error_values: [], colors: Auto, emit_json: false, emit_diffs: false, test_tool: Cargo }
DEBUG get_command_output{argv=["/home/kevin/.rustup/toolchains/nightly-x86_64-unknown-linux-gnu/bin/cargo", "locate-project"]}: output: {"root":"/home/kevin/polkadot-sdk/cumulus/parachains/integration-tests/emulated/tests/bridges/bridge-hub-rococo/Cargo.toml"}
DEBUG Found workspace root manifest cargo_toml_path="/home/kevin/polkadot-sdk/cumulus/parachains/integration-tests/emulated/tests/bridges/bridge-hub-rococo/Cargo.toml"
DEBUG resolved auto package filter to "bridge-hub-rococo-integration-tests"
DEBUG package{name=bridge-hub-rococo-integration-tests}: walk package manifest_path=/home/kevin/polkadot-sdk/cumulus/parachains/integration-tests/emulated/tests/bridges/bridge-hub-rococo/Cargo.toml
DEBUG package{name=bridge-hub-rococo-integration-tests}: found mutation target cumulus/parachains/integration-tests/emulated/tests/bridges/bridge-hub-rococo/src/lib.rs of kind ["lib"]
DEBUG source_file{path="cumulus/parachains/integration-tests/emulated/tests/bridges/bridge-hub-rococo/src/lib.rs"}: visit source file
TRACE source_file{path="cumulus/parachains/integration-tests/emulated/tests/bridges/bridge-hub-rococo/src/lib.rs"}:mod{line=17 mod_name="imports"}: mod excluded by attrs
TRACE source_file{path="cumulus/parachains/integration-tests/emulated/tests/bridges/bridge-hub-rococo/src/lib.rs"}:mod{line=77 mod_name="tests"}: mod excluded by attrs
```


You might be able to reproduce my issue by cloning [https://github.com/paritytech/polkadot-sdk](https://github.com/paritytech/polkadot-sdk), then `cd polkadot-sdk/cumulus/parachains/integration-tests/emulated/tests/bridges/bridge-hub-rococo`

**Reply by sourcefrog** [reply]
That crate has only one source file <https://github.com/paritytech/polkadot-sdk/blob/master/bridges/chains/chain-bridge-hub-rococo/src/lib.rs> ~~and it contains no functions, so there is nothing for cargo-mutants to mutate~~. Try running it in a crate that has more code.

Actually I spoke a bit too soon: there are a couple of functions that should be candidates:

https://github.com/paritytech/polkadot-sdk/blob/739951991f14279a7dc05d42c29ccf57d3740a4c/bridges/chains/chain-bridge-hub-rococo/src/lib.rs#L52-L61

```rust
	fn max_extrinsic_size() -> u32 {
		*BlockLength::get().max.get(DispatchClass::Normal)
	}

	fn max_extrinsic_weight() -> Weight {
		BlockWeightsForAsyncBacking::get()
			.get(DispatchClass::Normal)
			.max_extrinsic
			.unwrap_or(Weight::MAX)
	}
```

I'll have a look into why they're not being mutated. But the general advice still holds: run it in a place where there's more for it to do.

Possibly the macro there generates a lot of code, but cargo-mutants can't see it: https://mutants.rs/limitations.html?highlight=macro#cases-where-cargo-mutants--cant--help

**Reply by sourcefrog** [reply]
>Interestingly, it doesn't show the files under tests/

Yep, I can see how that would be confusing. But it's looking for files to mutate, and it mutates the code under test, not the tests. (Mutating tests might be interesting; there's a thread about it elsewhere, but it's not core at this time.)

**Reply by sourcefrog** [reply, marked as answer]
Actually my link there was to a different similarly-named crate.

You're running it in https://github.com/paritytech/polkadot-sdk/blob/master/cumulus/parachains/integration-tests/emulated/tests/bridges/bridge-hub-rococo/src/lib.rs.

All of the code in this crate is `#cfg[test]` and also there are no functions in it, so there's nothing to mutate. Try running it in a crate that has some non-test code.

**Reply by kevin-valerio** [reply]
Thank you very much for your help! I understand better now. I initially thought that the mutations would happen in the tests, and not in the code logic itself.

I wanted to test only some specific tests, without knowing which part of the code they are touching, that was my initial need. I'll investigate further.

Thanks again for your time :D

**Reply by sourcefrog** [reply]
Thanks! 

Added this in https://github.com/sourcefrog/cargo-mutants/wiki/Troubleshooting
